lib/node_modules/npm/bin/npm-cli.js

```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env node
+
+// IntelliJ/WebStorm needs a npm package bundled with a particular node installation.
+// Bundled npm package is detected as "/path/to/node/../lib/node_modules/npm/".
+// For "$(nodenv root)/shims/node", it's "$(nodenv root)/lib/node_modules/npm/".
+
+// To run a npm command, IntelliJ/WebStorm runs the main entry JavaScript file directly (npm/bin/npm-cli.js). Benefits:
+// * It's possible to use a selected interpreter (not from $PATH), preventing possible issues when 'node' is
+//   not found in $PATH due to configuration issues.
+// * It allows to pass node options (--debug/v8-options) if needed. Though none are currently passed.
+
+const nodenvRoot = require('path').resolve(__dirname, '../../../..')
```

Comment:
We shouldn't use `const` because it's not _fully_ supported until node 6. (Though this simple usage would probably work fine, at least to 0.10.)

Reply:
@jasonkarns Do you need a fix for this? AFAIU, it will be available as a plugin, so the change probably goes to a different repo? Feel free to close this PR.

Reply:
>But hardcoding the location?

@jasonkarns Agree, the hardcoded location is not nice as it requires to maintain magic filepath, `bin/npm-cli.js` in this case. Well, there hasn't been cases like this before when a bundled npm package couldn't be found by path to Node.js interpreter. So it worked well so far. In this case a clearer way to integrate would be to rely on `bin` property. Will be fixed in 2019.1 (https://youtrack.jetbrains.com/issue/WEB-36645).

>And having the shim get added to the user's PATH?

Do you mean `lib/node_modules/npm/bin` should be added to user's `PATH`? It worked for me locally without adding new locations to the `PATH`.

>So this will likely stay as a manual fix users of webstorm will have to do themselves.

Will it work for you if a package.json with `bin` property pointing to `bin/npm-cli.js` is added along with `bin/npm-cli.js`. Once IntelliJ/WebStorm supports `bin` property from package.json, JavaScript file location could be changed if there is a need for this.
Just not sure nodenv&IntelliJ/WebStorm users should suffer, because of hardcoded location which will be changed unlikely.

Regarding a package with that shim, IIUC, such a plugin was intended to be installed additionally to nodenv installation, so it won't work out-of-the-box?

Reply:
> Do you mean lib/node_modules/npm/bin should be added to user's PATH?

No! If we create a package named `jetbrains-npm-shim` that has `bin/npm-cli.js`, and we specify it's bin via: `"bin": "./bin/npm-cli.js"`, then when the user installs this package via npm or yarn, then they will have `jetbrains-npm-shim` in their `PATH`. That's how the `bin` property works. It tells npm which files to symlink into `PATH`!

Since webstorm _doesn't_ use `PATH` to find the npm script, then there's no need for it to be in `PATH` at all. But having this shim get added to `PATH` would be a negative side effect for webstorm user's who need to use the shim package.

Reply:
Actually, WebStorm uses `PATH` to find npm executable as an intermediate step to locate npm package directory. The problem is that the found npm executable script might be not a JavaScript file, e.g. it's so for Windows with npm.cmd. Initially, WebStorm tries to locate a npm package directory: first relative to the specified node interpreter `/path/to/node/../lib/node_modules/npm/`, then relative to the npm executable script found in the PATH `/path/to/npm/../lib/node_modules/npm/` (where `/path/to/npm` is an executable file), then, if still not found, it runs `readlink /path/to/npm` and assumes it's deferenced inside npm package. With the current nodenv approach (perfect for shell usage) `$(nodenv root)/shim/npm` doesn't link into a npm package and there is no `$(nodenv root)/lib/node_modules/npm`. So, a npm package cannot be discovered unfortunately.

If a shim package is named as `jetbrains-npm-shim` and its package.json contains `"bin": "./bin/npm-cli.js"`, then `jetbrains-npm-shim` executable will be in the system PATH. How was it supposed to be used? To search for `jetbrains-npm-shim` executable if `npm` is not found?

Reply:
So it looks like the `/path/to/node/../lib/node_modules/npm/` still requires configuring the package manager in webstorm?

With the following:

```
$ pwd
/usr/local/var/nodenv
$ ls shims/node
shims/node*
$ tree lib/
lib/
└── node_modules
    └── npm
        ├── bin
        │   └── npm-cli.js
        └── package.json

3 directories, 2 files
```

and the node interpreter set as: `/usr/local/var/nodenv/shims/node`

It triggers the popup to configure the project manager:

![image](https://user-images.githubusercontent.com/119972/50912304-d86b2700-13ff-11e9-9704-de393eb0e15e.png)

Now, I can see that it has derived the path to npm and is using the npm shim correctly. But one still has to select it? (which makes sense, I guess, to force the user to choose between npm or yarn?)

Reply:
The suggested managers list looks like:
![image](https://user-images.githubusercontent.com/119972/50912453-2aac4800-1400-11e9-9e88-60d0e091cad6.png)

Reply:
@jasonkarns Sorry for the delay. Text `Project` in "Package manager" field of the Npm run/debug configuration means that the package manager specified in "Settings | Languages & Frameworks | Node.js and NPM" should be used. Please check how "Package manager" is specified in "Settings | ... | Node.js and NPM". Looks like it's empty there.

Reply:
Right. I thought I read somewhere that `../lib/node_modules/npm/bin/npm-cli.js` would be picked up _automatically_ if it existed relative to the node interpreter. It was definitely _found_ by webstorm automatically (I can see it resolved the path and version number in the drop-down list), but I still had to explicitly select it when attempting to run a script.

I was thinking that it would be used automatically if found, without need to explicitly select it from the list.

Reply:
@jasonkarns According to your screenshots, the "npm" reference was resolved correctly to `/usr/local/var/nodenv/lib/node_modules/npm`. But the "Project" reference is unresolved unexpectedly. Could you please attach a screenshot of "Settings | Languages & Frameworks | Node.js and NPM" with expanded "Package manager" field?
Looks like a bug. Your file structure is still like described in https://github.com/nodenv/nodenv/pull/129#discussion_r246445026, right?

Reply:
@segrey I cleared it out manually hoping to simulate a "fresh install". apparently I just needed to restart webstorm and it selected that one automatically. all good 👍